test/lib/jdk/test/lib/util/ForceGC.java

```diff
@@ -78,9 +78,8 @@ public static boolean waitFor(BooleanSupplier booleanSupplier, long timeout) {
         ReferenceQueue<Object> queue = new ReferenceQueue<>();
         Object obj = new Object();
         PhantomReference<Object> ref = new PhantomReference<>(obj, queue);
-        obj = null;
         Reference.reachabilityFence(obj);
-        Reference.reachabilityFence(ref);
+        obj = null;
```

Comment:
You're right to question the utility of calling reachabilityFence(obj) after obj has been nulled out. But I'm still questioning the utility of calling RF(obj) at all. We don't care when obj is determined to be unreachable; what we care about is that the GC has done some reference processing. Seems to me we can simplify the above lines to

    PhantomReference<Object> ref = new PhantomReference<>(new Object(), queue);

and get rid of the local variable obj entirely.

Reply:
The reason for the explicit reference and RF, as I recall, is to guard against the allocation of the new object being elided entirely, with the `PhantomReference` constructor being passed null (or itself being elided) and no reference processing ever actually happening.

Reply:
What David says ;-)

Reply:
@dholmes-ora Is this really possible? The `obj` ref is passed to the PhantomReference constructor, which stores it in a field, the constructed PhantomReference is returned, and it's then used in a reachabilityFence call below. So `obj` should remain reachable the entire time, right?

Reply:
(As an aside, I wasn't able to determine what any of the Reference classes do if they're created with a null reference. Possibly spec bug?)

Reply:
@stuart-marks  My recollection, which I can't confirm is that this pattern was discussed internally and there was a lot of uncertainty about what was actually needed. Interestingly there was zero discussion of this in the actual PR that added it - https://github.com/openjdk/jdk/pull/8979

Thinking it through now, I tend to agree with you that the RF for `ref` suffices to prevent `obj` from being elided